op-node/rollup/event/limiter.go

```diff
@@ -0,0 +1,58 @@
+package event
+
+import (
+	"context"
+
+	"golang.org/x/time/rate"
+)
+
+type Limiter[E Emitter] struct {
+	ctx       context.Context
+	emitter   E
+	rl        *rate.Limiter
+	onLimited func()
+}
+
+// NewLimiter returns an event rate-limiter.
+// This can be used to prevent event loops from (accidentally) running too hot.
+// The eventRate is the number of events per second.
+// The eventBurst is the margin of events to eat into until the rate-limit kicks in.
+// The onLimited function is optional, and will be called if an emitted event is getting rate-limited
+func NewLimiter[E Emitter](ctx context.Context, em E, eventRate rate.Limit, eventBurst int, onLimited func()) *Limiter[E] {
+	return &Limiter[E]{
+		ctx:       ctx,
+		emitter:   em,
+		rl:        rate.NewLimiter(eventRate, eventBurst),
+		onLimited: onLimited,
+	}
+}
+
+// Emit is thread-safe, multiple parallel derivers can safely emit events to it.
+func (l *Limiter[E]) Emit(ev Event) {
+	if l.onLimited != nil && l.rl.Tokens() < 1.0 {
+		l.onLimited()
+	}
+	if err := l.rl.Wait(l.ctx); err != nil {
+		return // ctx error, safe to ignore.
```

Comment:
I'm a bit concerned: what if some critical event was dropped silently here?

Reply:
No, the Wait function error is limited to just the ctx. And all important events are processed before context cancellation. Besides, we do aim for events to be possible to drop without critical effects.